crates/shim-runc/src/container.rs

```diff
@@ -0,0 +1,366 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+#![allow(unused)]
+
+use std::collections::HashMap;
+use std::fs::{self, File};
+use std::io::{self, BufRead, BufReader, BufWriter, Write};
+use std::os::unix::fs::OpenOptionsExt;
+use std::path::{Path, PathBuf};
+use std::sync::{Arc, Mutex};
+
+use containerd_shim_protos as protos;
+
+use log::error;
+use protos::shim::shim::{DeleteRequest, KillRequest, StartRequest};
+
+use cgroups_rs as cgroup;
+
+use cgroup::Hierarchy;
+use nix::errno::Errno;
+use nix::sys::stat;
+use nix::unistd;
+use protobuf::Message;
+use sys_mount::UnmountFlags;
+use time::OffsetDateTime;
+
+use crate::options::oci::Options;
+use crate::process::{
+    config::{CreateConfig, MountConfig},
+    init::InitProcess,
+};
+use crate::utils;
+
+const OPTIONS_FILENAME: &str = "options.json";
+
+#[derive(Debug)]
+/// Struct for managing runc containers.
+pub struct Container {
+    mu: Arc<Mutex<()>>,
```

Comment:
Why mu is needed? The rust borrow checker should have already protected `Container` objects from concurrent access.